pkg/test/framework/components/echo/common/deployment/apps.go

```diff
@@ -0,0 +1,440 @@
+	Ns1 Namespace
+
+	// Echo deployments in the second namespace.
+	Ns2 Namespace
+
+	// Ns1AndNs2 combines all deployments in both Ns1 and Ns2.
+	Ns1AndNs2 echo.Services
+
+	// External (out-of-mesh) deployments
+	External External
+
+	// All echo instances.
+	All echo.Services
+}
+
+func (d *EchoDeployments) loadValues(t resource.Context, echos echo.Instances) error {
+	d.All = echos.Services()
+
+	if err := d.Ns1.loadValues(t, echos, d.Ns2.Namespace, d.External.Namespace); err != nil {
+		return err
+	}
+	if err := d.Ns2.loadValues(t, echos, d.Ns1.Namespace, d.External.Namespace); err != nil {
+		return err
+	}
+
+	d.Ns1AndNs2 = d.Ns1.All.Instances().Append(d.Ns2.All.Instances()).Services()
+
+	d.External.loadValues(echos)
+	return nil
+}
+
+const (
+	PodASvc          = "a"
+	PodBSvc          = "b"
+	PodCSvc          = "c"
+	PodTproxySvc     = "tproxy"
+	VMSvc            = "vm"
+	HeadlessSvc      = "headless"
+	StatefulSetSvc   = "statefulset"
+	ProxylessGRPCSvc = "proxyless-grpc"
+	NakedSvc         = "naked"
+	ExternalSvc      = "external"
+	DeltaSvc         = "delta"
+
+	externalHostname = "fake.external.com"
+)
+
+func serviceEntryPorts() []echo.Port {
+	var res []echo.Port
+	for _, p := range ports.All().GetServicePorts() {
+		if strings.HasPrefix(p.Name, "auto") {
+			// The protocol needs to be set in common.EchoPorts to configure the echo deployment
+			// But for service entry, we want to ensure we set it to "" which will use sniffing
+			p.Protocol = ""
+		}
+		res = append(res, p)
+	}
+	return res
+}
+
+func SetupApps(t resource.Context, i istio.Instance, apps *EchoDeployments) error {
+	var err error
+	apps.Ns1.Namespace, err = namespace.New(t, namespace.Config{
+		Prefix: "echo1",
+		Inject: true,
+	})
+	if err != nil {
+		return err
+	}
+	apps.Ns2.Namespace, err = namespace.New(t, namespace.Config{
+		Prefix: "echo2",
+		Inject: true,
+	})
+	if err != nil {
+		return err
+	}
+	apps.External.Namespace, err = namespace.New(t, namespace.Config{
+		Prefix: "external",
+		Inject: false,
+	})
+	if err != nil {
+		return err
+	}
+
+	apps.Ingress = i.IngressFor(t.Clusters().Default())
+	apps.Ingresses = i.Ingresses()
+
+	// Headless services don't work with targetPort, set to same port
+	headlessPorts := make([]echo.Port, len(ports.All()))
+	for i, p := range ports.All() {
+		p.ServicePort = p.WorkloadPort
+		headlessPorts[i] = p
+	}
+	builder := deployment.New(t).WithClusters(t.Clusters()...)
+	builder = buildNamespace(t, builder, apps.Ns1.Namespace)
+	builder = buildNamespace(t, builder, apps.Ns2.Namespace)
+	builder = buildExternal(builder, apps.External.Namespace)
+
+	echos, err := builder.Build()
+	if err != nil {
+		return err
+	}
+
+	// Load values from the deployed echo instances.
+	return apps.loadValues(t, echos)
+}
+
+// TODO: should t.Settings().Skip(echo.Delta) do all of this?
+func skipDeltaXDS(t resource.Context) bool {
+	return t.Settings().Skip(echo.Delta) || !t.Settings().Revisions.AtLeast("1.11")
```

Comment:
```suggestion
	return t.Settings().Skip(echo.Delta) || !t.Settings().Revisions.AtLeast("1.12")
```

changed in master

Reply:
Addressed by https://github.com/istio/istio/pull/38202